### ISSUE TITLE
Possibility to update a nested object's property in a non-mutable way

### ISSUE THREAD

**Issue by StudioSpindle** [jashkenas/underscore#2948]
Currently I am using something like:

```typescript
const foo = bar.map((baz) => {
  return {
    ...baz,
    qux: baz.qux.map((quux) => {
       return {
         ...quux,
         aNewProperty: 'yay!'
       }
    }
  }
});
```
To make the point clear, it would be ideal to have something similar to the [set method in lodash](https://lodash.com/docs/4.17.15#set):

```typescript
const foo = bar;
set(foo, 'bar.baz.qux.quux', { bar.baz.qux.quux, aNewPorperty: 'yay!' });
```

Note: I am not aware of how lodash and underscore are related but the current project I'm working on is using underscore. And at this time it's not possible to convert to lodash just for this single use-case.

**Comment by dogbubu** [comment]
gsdg54sdg5s4d5gドっぱくラバニラサダウのブはとてもおいしそうなりんかマカョ♂

**Comment by jgonggrijp** [comment]
Yes, I agree it would make sense to have a `set` function in Underscore as well. In the meanwhile, you can define your own `set` function like this so it will be maintainable and reusable and integrate well with Underscore:

```js
import _, { extend, isObject } from 'underscore';

var arrayIndex = /^\d+/;

function keyValue(key, value) {
    var result = {};
    result[key] = value;
    return result;
}

function innerSet(obj, path, value) {
    if (!path.length) return value;
    var key = path[0];
    // Important: the next line prevents prototype pollution.
    if (key === '__proto__') throw new Error('Prototype assignment attempted');
    obj = obj || (arrayIndex.test(key) ? [] : {}); 
    value = innerSet(obj[key], path.slice(1), value);
    return extend(obj, keyValue(key, value));
}

function set(collection, path, value) {
    if (!isObject(collection)) return collection;
    path = _.toPath(path);
    return innerSet(collection, path, value);
}
```

Use it like this:

```js
set(foo, ['bar', 'baz', 'qux', 'quux', 'aNewProperty'], 'yay!');
```

You can add it to Underscore so you can also use it in chaining:

```js
import { mixin } from 'underscore';
import { set } from './your/own/module.js';

_.mixin({set});

_.chain({a: 1})
.extend({b: 2})
.set(['c', 0, 'd'], 3)
.value();
// {a: 1, b: 2, c: [{d: 3}]}
```

If you want to use shorthand dotted paths of the form `'bar.baz.qux.quux.aNewProperty'`, you can enable this by overriding [`_.toPath`](https://underscorejs.org/#toPath), as long as you keep in mind the warning that comes with that.

> Note: I am not aware of how lodash and underscore are related but the current project I'm working on is using underscore. And at this time it's not possible to convert to lodash just for this single use-case.

Lodash is a fork of Underscore. Underscore is being actively maintained, so sticking with Underscore is fine.

**Comment by jgonggrijp** [comment]
I just noticed the "non-mutable" part of the issue title... sorry for missing that previously.

I want to avoid introducing new functions to Underscore that have the same name as a function in Lodash, but different semantics. So a `set` function in Underscore should be mutating. However, a `setClone` function is conceivable using `_.clone`, analogous to the code I wrote before:

```js
import { clone } from 'underscore';

function innerSetClone(obj, path, value) {
    if (!path.length) return value;
    var key = path[0];
    // Important: the next line prevents prototype pollution.
    if (key === '__proto__') throw new Error('Prototype assignment attempted');
    obj = obj || (arrayIndex.test(key) ? [] : {}); 
    value = innerSetClone(obj[key], path.slice(1), value);
    return extend(clone(obj), keyValue(key, value));
}

function setClone(collection, path, value) {
    if (!isObject(collection)) return collection;
    path = _.toPath(path);
    return innerSetClone(collection, path, value);
}
```

It does exactly the same thing as `set`, except that it always returns a new object (or array) and never mutates existing objects.

**Comment by jgonggrijp** [comment]
@sofia-80 I think I answered your question in what I wrote about `_.setClone` just above. What am I missing?

**Comment by pricklymusic** [comment]
Implementing `_.setClone` would be a cleaner approach than a mutating `set` for this use case, especially if we leverage a recursive function that utilizes `_.clone` at each level of the object path. Would you prefer this implementation to handle array indices via numeric path segments, or should we strictly target object keys to maintain performance for this specific pattern?

**Comment by jgonggrijp** [comment]
@pricklymusic Array indices via numeric path segments, because that is the convention we follow everywhere. Why would strictly targeting object keys be better for performance?
